src/pages/Storefront/Cart.js

```diff
@@ -16,17 +19,87 @@ import {
-import { PageContainer } from 'components/storefront/PageContainer/PageContainer';
-import CartSkeleton from 'components/storefront/Cart/Layout/CartSkeleton';
+
+const calculateDiscount = (discountApplications, lineItems) => {
+  let discountVal = 0;
+  let shippingCost = 'TBD';
+  let discountType = '';
+  if (discountApplications.length > 0) {
+    const appliedDiscount = discountApplications[0];
+    discountType = appliedDiscount.targetType;
+    if (discountType === 'SHIPPING_LINE') {
+      shippingCost = 'FREE';
+    } else if (discountType === 'LINE_ITEM') {
+      discountVal = lineItems
+        .reduce((acc, lineItem) => {
+          return lineItem.discountAllocations
+            ? acc +
+                lineItem.discountAllocations.reduce((acc2, discAlloc) => {
+                  return acc2 + parseFloat(discAlloc.allocatedAmount.amount);
+                }, 0)
+            : acc;
+        }, 0)
+        .toFixed(2);
+    }
+  }
+  return { discountVal, shippingCost, discountType };
+};
+
+const DiscountMsg = ({ discountApplications, discountSuccess, removeCoupon }) => {
+  let msg;
+  if (discountSuccess === 'true') {
+    msg = (
+      <VStack alignItems="left">
+        <chakra.h4 display="inline" whiteSpace="nowrap" textStyle="lightCaption">
+          Success! Your code <b>{discountApplications[0].code}</b> has been applied.
+        </chakra.h4>
+        <chakra.h4 textDecoration="underline" cursor="pointer" textStyle="lightCaption" onClick={removeCoupon}>
+          REMOVE COUPON CODE
+        </chakra.h4>
+      </VStack>
+    );
+  } else if (discountSuccess === 'false') {
+    msg = (
+      <chakra.h4 color="brand.red" textStyle="lightCaption">
+        Sorry, that coupon code is not applicable.
+      </chakra.h4>
+    );
+  }
+  return <>{msg}</>;
+};
 
 const CartItems = ({ checkoutData }) => {
-  const { id: checkoutId, lineItems } = checkoutData;
+  const { id: checkoutId, lineItems, discountApplications } = checkoutData;
+  const { addDiscount, removeDiscount } = useShopify();
   const cartItemsCount = lineItems.reduce((acc, cur) => acc + cur.quantity, 0);
+  const [discountCode, setDiscountCode] = useState('');
+  const [discountSuccess, setDiscountSuccess] = useState('');
 
-  // TO DO: May remove coupon & just let Shopify handle
-  const applyCoupon = () => {
-    alert('To be implemented');
+  const onChangeCoupon = e => {
+    // Assuming that all discounts are strictly uppercase
+    setDiscountCode(e.target.value.toUpperCase());
+  };
+
+  const applyCoupon = async () => {
+    const res = await addDiscount(checkoutId, discountCode);
+
+    // TO DO: Proper error handling (catch error in reducer?)
+    if (res.discountApplications.length > 0) {
+      // Handling case where there is an existing coupon + user enters invalid coupon
+      if (discountCode !== res.discountApplications[0].code) {
+        removeCoupon();
+        setDiscountSuccess('false');
```

Comment:
This was such a janky solution that somewhat works, I think this would be a lot easier (and would require some refactoring on my part) depending on how we do error handling. Also, I'm pretty sure while testing this was inconsistent but I didn't want to sink too much time since I'll probably rewrite it